flink-staging/flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkShell.scala

```diff
@@ -61,99 +73,217 @@ object FlinkShell {
+      }
+
+      (repl, cluster)
+    } catch {
+      case e: IllegalArgumentException =>
+        println(s"Error: ${e.getMessage}")
+        sys.exit()
+    }
 
-      // start scala interpreter shell
+    val settings = new Settings()
+    settings.usejavacp.value = true
+    settings.Yreplsync.value = true
+
+    try {
       repl.foreach(_.process(settings))
     } finally {
       repl.foreach(_.closeInterpreter())
-      cluster.foreach(_.stop())
+      cluster match {
+        case Some(Left(miniCluster)) => miniCluster.stop()
+        case Some(Right(yarnCluster)) => yarnCluster.shutdown(false)
+        case _ =>
+      }
+    }
+
+    println(" good bye ..")
+  }
+
+  def deployNewYarnCluster(yarnConfig: YarnConfig) = {
+    val yarnClient = FlinkYarnSessionCli.getFlinkYarnClient
+
+    // use flink-dist.jar for scala shell
+    val jarPath = new Path("file://" +
+      s"${yarnClient.getClass.getProtectionDomain.getCodeSource.getLocation.getPath}")
+    yarnClient.setLocalJarPath(jarPath)
+
+    // load configuration
+    val confDirPath = CliFrontend.getConfigurationDirectoryFromEnv
+    val flinkConfiguration = GlobalConfiguration.getConfiguration
+    val confFile = new File(confDirPath + File.separator + "flink-conf.yaml")
+    val confPath = new Path(confFile.getAbsolutePath)
+    GlobalConfiguration.loadConfiguration(confDirPath)
+    yarnClient.setFlinkConfigurationObject(flinkConfiguration)
+    yarnClient.setConfigurationDirectory(confDirPath)
+    yarnClient.setConfigurationFilePath(confPath)
+
+    // number of task managers is required.
+    yarnConfig.containers match {
+      case Some(containers) => yarnClient.setTaskManagerCount(containers)
+      case None =>
+        throw new IllegalArgumentException("Number of taskmanagers must be specified.")
+    }
+
+    // set configuration from user input
+    yarnConfig.jobManagerMemory.foreach(yarnClient.setJobManagerMemory)
+    yarnConfig.name.foreach(yarnClient.setName)
+    yarnConfig.queue.foreach(yarnClient.setQueue)
+    yarnConfig.slots.foreach(yarnClient.setTaskManagerSlots)
+    yarnConfig.taskManagerMemory.foreach(yarnClient.setTaskManagerMemory)
+
+    // deploy
+    val cluster = yarnClient.deploy()
+    val address = cluster.getJobManagerAddress.getAddress.getHostAddress
+    val port = cluster.getJobManagerAddress.getPort
+    cluster.connectToCluster()
+
+    (address, port, Some(Right(cluster)))
+  }
+
+  def fetchDeployedYarnClusterInfo() = {
+    // load configuration
+    val globalConfig = GlobalConfiguration.getConfiguration
+    val defaultPropertiesLocation = System.getProperty("java.io.tmpdir")
+    val currentUser = System.getProperty("user.name")
+    val propertiesLocation = globalConfig.getString(
+      ConfigConstants.YARN_PROPERTIES_FILE_LOCATION, defaultPropertiesLocation)
+    val propertiesName = CliFrontend.YARN_PROPERTIES_FILE + currentUser
+    val propertiesFile = new File(propertiesLocation, propertiesName)
```

Comment:
Maybe we could unify the `propertiesName` generation somewhere. It is also implemented in `ExecutionEnvironment`. If it changes at one place, then one has to remember that it also has to be changed here. This will most definitely be forgotten.